src/codegen/encoding/mod.rs

```diff
@@ -1732,10 +1735,57 @@ pub(super) trait AbiEncoding {
 
     /// Returns if the we are packed encoding
     fn is_packed(&self) -> bool;
+
+    /// Encode "Panic(uint256)" error data.
+    fn const_error_panic(&self, code: PanicCode) -> Vec<u8> {
```

Comment:
Wouldn't they be encoded the same in borsh?

Reply:
It looks like so. My concern was that this feature is not implemented yet on Solana. Anchor does not decode anything if a function fails, so we return no values.

Reply:
What about moving these functions out of the trait? They do not use `self`.

Reply:
Then, `create_encoder` need not to be `pub(crate)` and we can decouple the encoder from `emit`.

Reply:
First, I did it exactly like that. I reconsidered.

I _want_ anyone to force using the encoder, even when encoding const data.

We must not allow _any_ encoding happening outside of `AbiEncoding` trait. Otherwise we create a perfect footgun.

By implementing this here, as a trait method, we force this. It's exactly what I want.

Reply:
I'm less concerned about creating and using a `AbiEncoding` instance in `emit`. Yes, when we pass in a CFG that gets modified in emit, this would be problematic. However, this is 1. hard to do anyways, 2. very easy to spot in a code reviews. Where as without having us to force using the encoder, it's 1. super easy to emit incorrected const ABI encoded values and 2. much harder to spot in code reviews.

Reply:
Another benefit of having this in the `AbiEncoding` trait is that once we have an optimization to optimize encoding of compile time constant values, it can converge the arguments to constants and then just use this.

Reply:
> Another benefit of having this in the `AbiEncoding` trait is that once we have an optimization to optimize encoding of compile time constant values, it can converge the arguments to constants and then just use this.

I need to quote yourself for this one:

> Then we just fix it before implementing these optimizations

Reply:
Yes that's true. We can't even test this for Solana. I'm going to remove the provided implementations and implement it in the SCALE version.

Reply:
> I need to quote yourself for this one:
>> Then we just fix it before implementing these optimizations 

Do you think when we want to implement this optimization the constant ABI encoding can't be a trait method? Why?

Reply:
We could have generic `fn const_encode(args: &[Expression]) -> Vec<u8>` or something like that. That would be generic to useful for all targets and it can be used for both `Panic`, `Error()` and custom errors.

Reply:
Yes, that goes into the direction what I envision for #1451

Reply:
> We could have generic `fn const_encode(args: &[Expression]) -> Vec<u8>` or something like that. That would be generic to useful for all targets and it can be used for both `Panic`, `Error()` and custom errors.

That sounds like a good idea.

Reply:
Done